### PR TITLE
Add AMD ROCm HIP

### PR DESCRIPTION
**Pull request by cgmb**
This is an initial draft of the build process for the HIP library. At present, I've only enabled support for `VERSION=5.2.3`. A few notes:

* I put comgr, roct-thunk-interface, rocr-runtime and hip into `/opt/compiler-explorer/libs/rocm/5.2.3`, but I could split them into `/opt/compiler-explorer/libs/comgr/5.2.3`, `/opt/compiler-explorer/libs/hsakmt/5.2.3`, `/opt/compiler-explorer/libs/hsa-runtime64/5.2.3`, and `/opt/compiler-explorer/libs/hip/5.2.3` if you prefer. *UPDATE: It's probably simpler just to leave them bundled. They're all tied to the compiler they're built with anyway.*

* There's a very tight coupling between the compiler and comgr. This means that you essentially must use clang-rocm-5.2.3 when building comgr for rocm-5.2.3. It's possible to use other versions, but comgr uses unstable LLVM interfaces so there's often minor patches required for that to work.

~This is a draft as it has not yet undergone testing (though it's closely based on build scripts that were tested). I also have to track down how this gets invoked.~ I've finished testing and it's no longer a draft.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -31,6 +31,14 @@ RUN apt update -y -q && apt upgrade -y -q && apt update -y -q && \
...
     gettext \
     ninja-build \
+    file \
+    perl \
+    xxd \
+    mesa-common-dev \
+    zlib1g-dev \
+    libelf-dev \
+    libdrm-dev \
+    libudev-dev \
```
Comment (cgmb) [inline Dockerfile:41]: The need for these build dependencies is the only reason why I didn't just add an `install_hip()` to https://github.com/compiler-explorer/infra/blob/main/update_compilers/install_libraries.sh

Though, I suppose that if I built them from source, I could just this PR and do everything in infra.

**Comment by cgmb** [comment]
I'm having trouble figuring out how `build/build-hip.sh` should be invoked. The other scripts in misc-builder seem to be for nightly builds of compilers, so they get listed in the daily section of https://github.com/compiler-explorer/compiler-workflows. However, this is a library and I'm only building releases...

I suppose maybe `install_libraries.sh` in infra could execute this build in a misc-builder docker container, then copy the resulting artifacts out?

File: Dockerfile (hunk 1 above)
Comment (cgmb) [reply Dockerfile:41]: Ah. The install_libraries.sh script is deprecated. So, the misc-builder is the way to go.

**Comment by cgmb** [comment]
@mattgodbolt, thanks for the help. I've tested these with my local compiler-explorer instance and they now function correctly for me.

**Comment by mattgodbolt** [comment]
Awesome! We'll take a look at getting them worted!

File: Dockerfile [hunk 2]
```diff
@@ -31,6 +31,14 @@ RUN apt update -y -q && apt upgrade -y -q && apt update -y -q && \
     lldb \
     gettext \
     ninja-build \
+    file \
+    perl \
+    xxd \
+    mesa-common-dev \
```
Comment (mattgodbolt) [inline Dockerfile:37]: so long as the runtime for these are not required to run the compilers?

**Comment by mattgodbolt** [comment]
What version(s) would you recommend we build?

File: Dockerfile (hunk 2 above)
Comment (cgmb) [reply Dockerfile:37]: I'm pretty sure they're just build dependencies. mesa-common-dev definitely is required after building.

perl and file are needed for hipcc, but I was going to just set up CE to use the ROCm clang directly, so they're not needed.

I'm pretty sure xxd is also just a build dependency. It was just needed to compile rocr.

**Comment by cgmb** [comment]
> What version(s) would you recommend we build?

This should be able to build `4.5.2`, `5.0.2`, `5.1.3`, and 5.2.3. The build process for HIP prior to ROCm 4.5 was different, so that's why I picked 4.5 as a cutoff point for old versions.

I basically just picked the last patch release of each minor version. The release cadence is roughly one minor version every two months. It's really more a question of how many binaries you feel like building and hosting.

File: Dockerfile (hunk 2 above)
Comment (cgmb) [reply Dockerfile:37]: I just did a couple more tests. perl is part of the ubuntu:20.04 base docker image, so I didn't try removing it, but my local copy of CE works fine without mesa, xdd or file. 

I can confirm that they're only needed at library build time. They are not needed for compiling HIP code.

Comment (mattgodbolt) [reply Dockerfile:37]: Fantastic! I'll try and get this all reviewed and tested tongith, but if not, Friday is my next concerted Compiler Explorer time.

**Comment by mattgodbolt** [comment]
Thanks will kick builds off shortly!

**Comment by mattgodbolt** [comment]
Builds:
- https://github.com/compiler-explorer/infra/actions/runs/3152679964
- https://github.com/compiler-explorer/infra/actions/runs/3152680934
- https://github.com/compiler-explorer/infra/actions/runs/3152682211
- https://github.com/compiler-explorer/infra/actions/runs/3152683390

**Comment by mattgodbolt** [comment]
All seem to have succeeded! yay

**Comment by cgmb** [comment]
> Builds:
> 
>     * https://github.com/compiler-explorer/infra/actions/runs/3152679964
> 
>     * https://github.com/compiler-explorer/infra/actions/runs/3152680934
> 
>     * https://github.com/compiler-explorer/infra/actions/runs/3152682211
> 
>     * https://github.com/compiler-explorer/infra/actions/runs/3152683390

@mattgodbolt, these will need to be re-run now that #48 is merged.

**Pull request merged**